File: back/src/entities/Snake.ts
import GameEntity, { GameEntityOptions } from './GameEntity';
import Food from './Food';
import config from '../config';
import randomcolor = require('randomcolor');
const simplify = require('simplify-js');

const { SPEED, BASE_ANGLE, INITIAL_SCALE, MAX_SIZE, MAX_LEN } = config;

interface SnakeOptions extends GameEntityOptions {
  id: string;
  length?: number;
  angle?: number;
  username: string;
  fillColor?: string;
  strokeColor?: string;
}

export class Movement {
  constructor(
    public x: number,
    public y: number,
    public speed: number,
    public angle: number,
  ) {}
}

export default class Snake extends GameEntity {
  public score: number = 0;
  public isSpeedUp: boolean = false;
  public fillColor: string = '';
  public angle: number;
  public stopped: boolean = false;
  public id: string;
  public scale: number = INITIAL_SCALE;
  public isBlinking: boolean = false;
  public collisionRect: {
    minX: number;
    minY: number;
    maxX: number;
    maxY: number;
  };

  public frameCounter: number = 0;

  // save snake's movement
  public movementQueue: Movement[] = [];
  public username: string;
  // max length of queue
  public movementQueueLen: number;
  public speed: number = SPEED;
  public oldSpeed: number = SPEED;
  public length: number;
  public toAngle: number;
  private turnSpeed: number = 0.15;
  private vx: number = 0;
  private vy: number = 0;

  public points: {
    x: number;
    y: number;
  }[] = [];

  public simplifiedPoints: {
    x: number;
    y: number;
  }[] = [];

  constructor(options?: SnakeOptions) {
    super(options);
    this.id = options.id;
    const strokeColor: string = options.strokeColor || '#000';
    this.fillColor = randomcolor();
    this.toAngle = this.angle = (options.angle || 0) + BASE_ANGLE;
    this.length = options.length;
    this.collisionRect = {
      minX: 30000,
      maxX: -30000,
      minY: 30000,
      maxY: -30000,
    };
    this.username = options.username;
    this.updateSize();
    this.velocity();
  }

  public updateSize(added: number = 0): void {
    this.width = Math.min(this.width + added,  MAX_SIZE);
    this.height = Math.min(this.height + added, MAX_SIZE);
    this.length += Math.min(added * 50, MAX_LEN);
    this.turnSpeed -= added / 1000;
    this.turnSpeed = Math.max(0.08, this.turnSpeed);
    this.movementQueueLen = Math.ceil(this.length / this.oldSpeed);
  }

  // move to new position
  public moveTo(nx: number, ny: number): void {
    const x: number = nx - this.x;
    const y: number = this.y - ny;
    let angle: number = Math.atan(Math.abs(x / y));

    // calculate angle, value is 0-360
    if (x > 0 && y < 0) {
      angle = Math.PI - angle;
    } else if (x < 0 && y < 0) {
      angle = Math.PI + angle;
    } else if (x < 0 && y > 0) {
      angle = Math.PI * 2 - angle;
    }

    const oldAngle: number = Math.abs(this.toAngle % (Math.PI * 2));

    // number of turns
    let rounds: number = ~~(this.toAngle / (Math.PI * 2));

    this.toAngle = angle;

    if (oldAngle >= (Math.PI * 3) / 2 && this.toAngle <= Math.PI / 2) {
      // move from fourth quadrant to first quadrant
      rounds += 1;
    } else if (oldAngle <= Math.PI / 2 && this.toAngle >= (Math.PI * 3) / 2) {
      // move from first quadrant to fourth quadrant
      rounds -= 1;
    }

    // calculate the real angle by rounds
    this.toAngle += rounds * Math.PI * 2;
  }

  // calculate horizontal speed and vertical speed by angle of snake header
  public velocity(): void {
    const angle: number = this.angle % (Math.PI * 2);
    const vx: number = Math.abs(this.speed * Math.sin(angle));
    const vy: number = Math.abs(this.speed * Math.cos(angle));

    if (angle < Math.PI / 2) {
      this.vx = vx;
      this.vy = -vy;
    } else if (angle < Math.PI) {
      this.vx = vx;
      this.vy = vy;
    } else if (angle < (Math.PI * 3) / 2) {
      this.vx = -vx;
      this.vy = vy;
    } else {
      this.vx = -vx;
      this.vy = -vy;
    }
  }

  // turn around
  public turnAround(): void {
    const angleDistance: number = this.toAngle - this.angle;

    if (Math.abs(angleDistance) <= this.turnSpeed) {
      // reset angle
      this.toAngle = this.angle = BASE_ANGLE + (this.toAngle % (Math.PI * 2));
    } else {
      this.angle += Math.sign(angleDistance) * this.turnSpeed;
    }
  }

  public speedUp(): void {
    if (this.isSpeedUp) {
      return;
    }

    this.isSpeedUp = true;
    this.oldSpeed = this.speed;
    this.speed *= 2;
  }

  public speedDown(): void {
    if (!this.isSpeedUp) {
      return;
    }

    this.isSpeedUp = false;
    this.speed = this.oldSpeed;
  }

  // eat food
  public eat(food: Food): number {
    this.score += food.value;

    this.frameCounter = 10;

    // add points
    const added = food.value / 200;
    this.updateSize(added);
    return added;
  }

  public blink(): void {
    this.isBlinking = !this.isBlinking;
  }

  public updateCollisionRect(): void {
    this.simplifiedPoints = simplify(this.points, 5);
    // this.simplifiedPoints = this.points.slice(9, this.points.len);
    this.collisionRect = {
      minX: 3000,
      maxX: -3000,
      minY: 3000,
      maxY: -3000,
    };

    this.simplifiedPoints.forEach((point: any) => {
      if (point.x < this.collisionRect.minX) {
        this.collisionRect.minX = point.x;
      }
      if (point.x > this.collisionRect.maxX) {
        this.collisionRect.maxX = point.x;
      }
      if (point.y < this.collisionRect.minY) {
        this.collisionRect.minY = point.y;
      }
      if (point.y > this.collisionRect.maxY) {
        this.collisionRect.maxY = point.y;
      }
    });

    this.collisionRect.minX -= this.width;

    this.collisionRect.minY -= this.width;

    this.collisionRect.maxX += this.width;

    this.collisionRect.maxY += this.width;
  }

  // snake action
  public action() {
    if (this.stopped) {
      return;
    }

    if (this.frameCounter > 0) {
      this.blink();
      this.frameCounter -= 1;
    } else {
      this.isBlinking = false;
    }

    if (this.isSpeedUp) {
      this.length = Math.max(this.length - 2, 0);
      this.movementQueueLen = Math.max(this.movementQueueLen - 2, 0);
    }

    // save movement
    this.movementQueue.push(
      new Movement(this.x, this.y, this.speed, this.angle),
    );

    if (this.movementQueue.length > this.movementQueueLen) {
      this.movementQueue.shift();
    }

    this.turnAround();
    this.velocity();
    this.x += this.vx;
    this.y += this.vy;

    this.points = [];
    let wholeLength = this.length;
    if (this.movementQueue.length) {
      let i = this.movementQueue.length - 1;
      while (i > 0) {
        const movement = this.movementQueue[i];
        let x = movement.x;
        let y = movement.y;
        if (wholeLength > 0 && wholeLength < movement.speed) {
          const lm = this.movementQueue[i + 1] || this;
          const ratio = wholeLength / movement.speed;
          x = lm.x - (lm.x - x) * ratio;
          y = lm.y - (lm.y - y) * ratio;
        } else if (wholeLength < 0) {
          break;
        }

        i -= 1;
        wholeLength -= movement.speed;
        this.points.push({ x, y });
      }
    }
    this.updateCollisionRect();
  }
}

File: back-go/snake.go
package main

import (
	"bytes"
	"math"
	"math/rand"

	"github.com/teris-io/shortid"
)

type Point struct {
	X float64 `msgpack:"x"`
	Y float64 `msgpack:"y"`
}

type Movement struct {
	Point
	Speed float64
	Angle float64
}

type Rect struct {
	MinX float64 `msgpack:"minX"`
	MinY float64 `msgpack:"minY"`
	MaxX float64 `msgpack:"maxX"`
	MaxY float64 `msgpack:"maxY"`
}

type GameEntity struct {
	Point
	Width         float64 `msgpack:"width"`
	Height        float64 `msgpack:"height"`
	CollisionRect *Rect   `msgpack:"collisionRect"`
}

type Snake struct {
	GameEntity
	Id               string  `msgpack:"id"`
	Score            int     `msgpack:"score"`
	IsSpeedUp        bool    `msgpack:"isSpeedUp"`
	FillColor        string  `msgpack:"fillColor"`
	Angle            float64 `msgpack:"angle"`
	Scale            float64 `msgpack:"scale"`
	IsBlinking       bool    `msgpack:"isBlinking"`
	Username         string  `msgpack:"username"`
	Speed            float64 `msgpack:"speed"`
	Length           float64 `msgpack:"length"`
	DisplayRect      *Rect
	FrameCounter     int
	MovementQueue    []*Movement
	OldSpeed         float64
	Stopped          bool
	ToAngle          float64
	TurnSpeed        float64
	Points           []*Point
	SimplifiedPoints []*Point
	MovementQueueLen float64
	Vx               float64
	Vy               float64
}

func getRandomColor() string {
	letters := []byte{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'}
	color := bytes.NewBufferString("#")
	for i := 0; i < 6; i++ {
		color.WriteByte(letters[rand.Intn(16)])
	}
	return color.String()
}

func NewSnake(username string) *Snake {
	r := rand.Float64()
	s := &Snake{
		GameEntity: GameEntity{
			Point: Point{
				X: math.Floor(rand.Float64()*(MAP_WIDTH-100) + 100/2),
				Y: math.Floor(rand.Float64()*(MAP_HEIGHT-100) + 100/2),
			},
			Width:  30,
			Height: 30,
			CollisionRect: &Rect{
				MinX: 30000,
				MaxX: -30000,
				MinY: 30000,
				MaxY: -30000,
			},
		},
		Id:               shortid.MustGenerate(),
		FillColor:        getRandomColor(),
		Angle:            r*math.Pi*2 + BASE_ANGLE,
		ToAngle:          r*math.Pi*2 + BASE_ANGLE,
		Length:           280,
		Username:         username,
		Scale:            INITIAL_SCALE,
		Speed:            SPEED,
		OldSpeed:         SPEED,
		TurnSpeed:        0.15,
		MovementQueueLen: math.Floor(280 / SPEED),
	}
	s.UpdateSize(0)
	s.Velocity()
	return s
}

func (s *Snake) UpdateSize(added float64) {
	s.Width += added
	s.Height += added
	s.Length += added * 50
	s.TurnSpeed -= added / 1000
	s.TurnSpeed = math.Max(0.05, s.TurnSpeed)
	s.MovementQueueLen = math.Floor(float64(s.Length) / s.OldSpeed)
}

func (s *Snake) MoveTo(nx, ny float64) {
	x := nx - s.X
	y := s.Y - ny
	angle := math.Atan(math.Abs(x / y))

	if x > 0 && y < 0 {
		angle = math.Pi - angle
	} else if x < 0 && y < 0 {
		angle = math.Pi + angle
	} else if x < 0 && y > 0 {
		angle = math.Pi*2 - angle
	}

	oldAngle := math.Abs(math.Mod(s.ToAngle, (math.Pi * 2)))

	rounds := math.Floor(s.ToAngle / (math.Pi * 2))

	s.ToAngle = angle

	if oldAngle >= (math.Pi*3)/2 && s.ToAngle <= math.Pi/2 {
		rounds += 1
	} else if oldAngle <= math.Pi/2 && s.ToAngle >= (math.Pi*3)/2 {
		rounds -= 1
	}

	s.ToAngle += rounds * math.Pi * 2
}

func (s *Snake) Velocity() {
	angle := math.Mod(s.Angle, math.Pi*2)
	vx := math.Abs(s.Speed * math.Sin(angle))
	vy := math.Abs(s.Speed * math.Cos(angle))

	if angle < math.Pi/2 {
		s.Vx = vx
		s.Vy = -vy
	} else if angle < math.Pi {
		s.Vx = vx
		s.Vy = vy
	} else if angle < (math.Pi*3)/2 {
		s.Vx = -vx
		s.Vy = vy
	} else {
		s.Vx = -vx
		s.Vy = -vy
	}
}

func Sign(x float64) float64 {
	if x > 0 {
		return 1
	}
	if x < 0 {
		return -1
	}
	return 0
}

func (s *Snake) TurnAround() {
	angleDistance := s.ToAngle - s.Angle

	if math.Abs(angleDistance) <= s.TurnSpeed {
		s.Angle = BASE_ANGLE + math.Mod(s.ToAngle, math.Pi*2)
		s.ToAngle = s.Angle
	} else {
		s.Angle += Sign(angleDistance) * s.TurnSpeed
	}
}

func (s *Snake) SpeedUp() {
	if s.IsSpeedUp {
		return
	}

	s.IsSpeedUp = true
	s.OldSpeed = s.Speed
	s.Speed *= 2
}

func (s *Snake) SpeedDown() {
	if !s.IsSpeedUp {
		return
	}

	s.IsSpeedUp = false
	s.Speed = s.OldSpeed
}

func (s *Snake) Eat(f *Food) float64 {
	s.Score += int(f.Value)

	s.FrameCounter = 10

	added := f.Value / 200
	s.UpdateSize(added)
	return added
}

func (s *Snake) Blink() {
	s.IsBlinking = !s.IsBlinking
}

func (s *Snake) UpdateCollisionRect() {
	s.SimplifiedPoints = Simplify(s.Points, 5, false)

	s.CollisionRect = &Rect{
		MinX: 3000,
		MaxX: -3000,
		MinY: 3000,
		MaxY: -3000,
	}

	for _, point := range s.SimplifiedPoints {
		if point.X < s.CollisionRect.MinX {
			s.CollisionRect.MinX = point.X
		}
		if point.X > s.CollisionRect.MaxX {
			s.CollisionRect.MaxX = point.X
		}
		if point.Y < s.CollisionRect.MinY {
			s.CollisionRect.MinY = point.Y
		}
		if point.Y > s.CollisionRect.MaxY {
			s.CollisionRect.MaxY = point.Y
		}
	}

	s.CollisionRect.MinX -= s.Width
	s.CollisionRect.MinY -= s.Width
	s.CollisionRect.MaxX += s.Width
	s.CollisionRect.MaxY += s.Width

	s.DisplayRect = &Rect{
		s.CollisionRect.MinX - 950,
		s.CollisionRect.MinY - 450,
		s.CollisionRect.MaxX + 950,
		s.CollisionRect.MaxY + 450,
	}
}

func (s *Snake) Action() {
	if s.Stopped {
		return
	}

	if s.FrameCounter > 0 {
		s.Blink()
		s.FrameCounter -= 1
	} else {
		s.IsBlinking = false
	}

	if s.IsSpeedUp {
		s.Length -= 2
		s.MovementQueueLen -= 2
	}

	s.MovementQueue = append(s.MovementQueue, &Movement{Point{s.X, s.Y}, s.Speed, s.Angle})

	if len(s.MovementQueue) > int(s.MovementQueueLen) {
		s.MovementQueue = s.MovementQueue[1:]
	}

	s.TurnAround()
	s.Velocity()
	s.X += s.Vx
	s.Y += s.Vy

	s.Points = []*Point{}
	wholeLength := s.Length

	mLength := len(s.MovementQueue)
	if mLength > 0 {
		i := mLength - 1
		for i > 0 {
			movement := s.MovementQueue[i]
			x := movement.X
			y := movement.Y
			if wholeLength > 0 && wholeLength < movement.Speed {
				var lm *Movement
				if i < len(s.MovementQueue)-1 {
					lm = s.MovementQueue[i+1]
				} else {
					lm = &Movement{Point{s.X, s.Y}, s.Width, s.Height}
				}
				ratio := wholeLength / movement.Speed
				x = lm.X - (lm.X-x)*ratio
				y = lm.Y - (lm.Y-y)*ratio
			} else if wholeLength < 0 {
				break
			}
			i--
			wholeLength -= movement.Speed
			s.Points = append(s.Points, &Point{x, y})
		}
	}

	s.UpdateCollisionRect()
}
